include/internal/hashtable.h

```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef OPENSSL_HASHTABLE_H
+# define OPENSSL_HASHTABLE_H
+# pragma once
+
+#include <stddef.h>
+#include <stdint.h>
+
+typedef struct ht_internal_st HT;
+
+/*
+ * Represents a value in the hash table
+ */
+typedef struct ht_value_st {
+    void *value;
+    uint32_t type_id;
+} HT_VALUE;
+
+/*
+ * Represents a list of values filtered from a hash table
+ */
+typedef struct ht_value_list_st {
+    size_t list_len;
+    HT_VALUE **list;
+} HT_VALUE_LIST;
+
+/*
+ * Hashtable configuration
+ */
+typedef struct ht_config_st {
+    void (*ht_free_fn)(HT_VALUE *obj);
+    uint64_t (*ht_hash_fn)(uint8_t *key, size_t keylen);
+    uint32_t init_neighborhoods;
+    int lockless_write;
+    int lockless_read;
+    int dont_refcount;
+} HT_CONFIG;
+
+/*
+ * Key value for a hash lookup
+ */
+typedef struct ht_key_header_st {
+    size_t keysize;
+    uint8_t *keybuf;
+} HT_KEY;
+
+/*
+ * Hashtable key rules
+ * Any struct can be used to formulate a hash table key, as long as the
+ * following rules
+ * 1) The first element of the struct defining the key must be an HT_KEY
```

Comment:
-1, this is actually less generic than LHASH and makes it problematic to have the same heap-allocated object in more than one hashtable.

Reply:
I disagree.  I did exactly what you are suggesting in the namemap, creating a second reverse lookup table for the num2name call, so as to avoid the need for a complete iteration of the table when mapping numbers to names.  Both tables stored the same data, the reverse table just didn't specify a free function.  This allows multiple hash tables to store the same data easily

Reply:
I'm sorry, I should clarify here.  What I did when testing the handshake test with the updated namemap was to instantiate a second datatype with 2 instances of the IMPLEMENT_HT_VALUE_TYPE_FNS macros, one for name 2 number mappings and one for number to name mappings.  The second type allows me to insert a reverse mapping using a number as a lookup, and the data type holds a chain of namenum entries that use that number.  The eliminates the need to do a foreach iteration when doing a call to ossl_namemap_doall_names.  Instead now we just do a single lookup against the number and traverse the chain for each callback.

I understand this isn't what you're saying above, but my point is that the same works in reverse.  I just constructed the free function for the hash table to not free the namenum entries when the type of data is NUMNAME_ENTRY.  The same can be done in reverse, if you want to to store the same instance in multiple hash tables.  Possible implementation could be deciding at design time which table owns the heap allocated object, and only freeing it when freed from that table.  Another (better) solution is to have the heap allocated object be ref-counted, and do a decrement-and free on zero in each hash tables free function.  If we wanted to get clever, we could extend the api to create a event callback between tables (i.e. notify a subordonate hash table when the "owning" hash table deletes an object shared between the two, so that the subordonate can clean up its entries as well, though that seems a little cumbersome.